### ISSUE TITLE
Feature Request/Idea: UI support for nested compound fields

### ISSUE THREAD

**Issue by vera** [IQSS/dataverse#9200]
**Overview of the Feature Request**

Currently, Dataverse doesn't fully support nested compound fields (i.e. compound fields within compound fields). More precisely, the UI cannot display them, although the API and the internal data model already appear to support them. In some constellations, the SOLR schema is also not updated correctly.

**What kind of user is the feature intended for?**

Any user using the UI to view/edit datasets (curator, depositor, guest, superuser...) and API users

**What inspired the request?**

We would like to use custom metadata schemas that contain nested compound fields.

DataCite v>=4.1 also contains fields like this, e.g. see "7. Contributor" (0-n) with its children "7.4 nameIdentifier" (0-n) and "7.5 affiliation" (0-n) in https://schema.datacite.org/meta/kernel-4.4/doc/DataCite-MetadataKernel_v4.4.pdf

I have tested the following three basic scenarios of nested compound fields (load custom TSV, upload dataset via API, export dataset via API):

1. compound field nested within compound field
2. like 1. but outer field has `allowmultiple TRUE`
3. like 1. but both fields have `allowmultiple TRUE`

The TSV files and JSON datasets are here: [nested_compound_test.tar.gz](https://github.com/IQSS/dataverse/files/10124812/nested_compound_test.tar.gz)

Although the UI cannot display the nested fields, in scenarios 1 and 3 the export actually matched the initial upload, thus it seems the API and the internal data model already support this feature. In scenario 2 I got an error message at export (`Exception thrown from bean: org.apache.solr.client.solrj.impl.HttpSolrClient$RemoteSolrException: Error from server at http://solr:8983/solr/collection1: ERROR: [doc=dataset_9_draft] multiple values encountered for non multiValued field`). After manually editing the SOLR schema file to set `multiValued="true"`, the export succeeded here as well.

**What existing behavior do you want changed?**

The SOLR schema should be updated correctly when a TSV with nested compound fields is uploaded.

**Any brand new behavior do you want to add to Dataverse?**

The UI should be able to display nested compound fields when creating, editing and viewing a dataset.

Currently, it looks like this (empty because inner children aren't shown):

![Screenshot 2022-11-30 at 17-15-55 Add New Dataset - Root](https://user-images.githubusercontent.com/16904069/204850931-6b6452c2-4529-411f-b0be-7112e8d950e6.png)

**Any related open or closed issues to this feature request?**

https://github.com/IQSS/dataverse/issues/377 is related. When allowing multiples of a compound child, in some cases the compound child itself should also be a compound, e.g.:

"Author" is a compound field. To allow an author to have multiple identifiers, the children "Identifier Scheme" and "Identifier" should be grouped in a repeatable nested compound field.

(See also "7. Contributor" (0-n), "7.4 nameIdentifier" (0-n), "7.4.a nameIdentifierScheme", "7.4.b schemeURI" in https://schema.datacite.org/meta/kernel-4.4/doc/DataCite-MetadataKernel_v4.4.pdf (present since DataCite v4.1 as property 7.5))

cc @johannes-darms

**Comment by poikilotherm** [comment]
For the Solr schema discussion, you might be interested in #5989 and #7662. 

I did some work for that before, see [here for 5989](https://github.com/poikilotherm/dataverse/tree/5989-refactor-solrfield) and [here for the newer approach on 7662](https://github.com/poikilotherm/dataverse/tree/7662-solrconfig) with an (unfinished) TSV block parser to validate blocks and create the complete Solr core template as a package.

**Comment by pdurbin** [comment]
@vera thanks for opening this issue! For now I'm just adding ".txt" to the files in your tarball so I can more easily look at them from this issue. I might update this comment in a bit with something more substantial to say. 😄 

- [nested_compound_test1.tsv.txt](https://github.com/IQSS/dataverse/files/10493059/nested_compound_test1.tsv.txt)
- [nested_compound_test1.json.txt](https://github.com/IQSS/dataverse/files/10493061/nested_compound_test1.json.txt)
- [nested_compound_test2.tsv.txt](https://github.com/IQSS/dataverse/files/10493060/nested_compound_test2.tsv.txt)
- [nested_compound_test2.json.txt](https://github.com/IQSS/dataverse/files/10493063/nested_compound_test2.json.txt)
- [nested_compound_test3.tsv.txt](https://github.com/IQSS/dataverse/files/10493064/nested_compound_test3.tsv.txt)
- [nested_compound_test3.json.txt](https://github.com/IQSS/dataverse/files/10493062/nested_compound_test3.json.txt)

**Comment by beepsoft** [comment]
We at [ARP](https://researchdata.hu/en) would be really interested in this as well. This is currently the only thing stopping us introducing some advanced schemas in our Dataverse installation. We already use [CEDAR](https://metadatacenter.org/) for authoring the Dataverse schemas/metadatblocks and CEDAR allows any depth of nesting but we cannot allow such schemas currently to be imported into Dataverse.

**Comment by johannes-darms** [comment]
@beepsoft Most of the feature is already implemented and working, the only issues are a bug in the OAI_ORE exporter and the missing recursive implementation in JSF for the form and displaying of metadata. So the feature should be pretty easy to implement.

**Comment by beepsoft** [comment]
Thanks for the update @johannes-darms! So, you mean the internals - backend - is ready for defining metadatablocks with any nesting of fields and storing metadata in any depth of DatasetFieldCompoundValues only the metadata UI needs update?

**Comment by johannes-darms** [comment]
Yes, that is the case. Yet, there are no test for this feature...

**Comment by beepsoft** [comment]
That sounds great!

**Comment by vera** [comment]
> bug in the OAI_ORE exporter

Adding a link to the related issue: https://github.com/IQSS/dataverse/issues/10809

**Comment by cmbz** [comment]
2025/01/27: Keeping, related to other support for nested compound fields.
